hermes/probe/monitor.go

```diff
@@ -12,9 +12,164 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
+// Authors:
+// - Alicja Kwiecinska (kwiecinskaa@google.com) github: alicjakwie
+// - Evan Spendlove, GitHub: evanSpendlove.
+//
 // Monitor implements a probe that monitors a storage system using the
 // Hermes algorithm.
+// Package probe implements the probe that Hermes uses to monitor
+// a storage system.
+
+package probe
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+	"github.com/googleinterns/step224-2020/hermes"
+
+	probepb "github.com/googleinterns/step224-2020/config/proto"
+	hermespb "github.com/googleinterns/step224-2020/hermes/proto"
+)
 
 // MonitorProbe holds aggregate information about all probe runs, per-target.
 // It also holds the config and options used to initialise the probe.
-type MonitorProbe struct{}
+type MonitorProbe struct {
+	name    string
+	Config  *probepb.HermesProbeDef
+	targets []*probepb.Target
+	opts    *options.Options
+
+	Hermes *hermes.Hermes
+
+	MetricMu sync.RWMutex
+	// latencyMetrics stores the api call and probe operation latency per target.
+	// Metrics are stored with additional labels to record operation type and exit status.
+	LatencyMetrics map[string]*probeRunMetrics
+
+	Logger *logger.Logger
+}
+
+// Init initializes the probe with the given parameters.
+// This is a required method to implement the cloudprober.Probes.Probe interface.
+func (mp *MonitorProbe) Init(name string, opts *options.Options) error {
+	conf, ok := opts.ProbeConf.(*probepb.HermesProbeDef)
+	if !ok {
+		return fmt.Errorf("invalid argument: opts.ProbeConf is not of type *probepb.HermesProbeDef")
+	}
+	mp.name = name
+	mp.Config = conf
+	mp.targets = mp.Config.GetTargets()
```

Comment:
You could optimize the storage for mp by not duplicating information it already has. Remove this line, and simply iterate/pass around mp.Config.GetTargets(). 

Alternatively, if you aren't using any other member field from mp.Config, you can just have MetricProbe take just take in an array of targets.

Reply:
I've removed this duplication due to refactoring.